app/rankers/algorithms.py

```diff
@@ -246,15 +248,41 @@ def user_impression_filter(recs: RecommendationListType, user_impressed_list: Li
 def rank_by_preferred_topics(
         recs: CorpusRecommendationListType,
         preferred_topics: List[TopicModel],
+        count: int
 ) -> CorpusRecommendationListType:
     """
-    Sorts recommendations in-order such that recs with a preferred topic rank above recs with a non-preferred topic.
-    :param recs:
+    Recommends items with preferred topics, rotating the topics. Items for each topic are ordered randomly.
+    In case there's not enough items for the preferred topics, it adds the rest of the candidates in the provided order.
+
+    :param recs: candidates
     :param preferred_topics: List of topics that the user has expressed a preference for.
-    :return: ordered list of recommendations with preferred topics ranking higher.
+    :param count: a number of items to recommend
+    :return: ordered list of recommendations with preferred topics.
     """
-    preferred_corpus_topic_ids = {topic.corpus_topic_id for topic in preferred_topics}
-    return sorted(recs, key=lambda rec: rec.topic in preferred_corpus_topic_ids, reverse=True)
+    topic_ids = {topic.corpus_topic_id for topic in preferred_topics}
+    recs_by_topic = defaultdict(list)
+    for rec in recs:
+        if rec.topic in topic_ids:
+            recs_by_topic[rec.topic].append(rec)
+
+    res = []
+    recommended = set()
+    while len(res) < count and len(recs_by_topic) > 0:
+        for topic in topic_ids:
+            if topic not in recs_by_topic:
+                continue
+            topic_recs = recs_by_topic[topic]
+            position = random.randint(0, len(topic_recs) - 1) if len(topic_recs) > 0 else 0
+            item = topic_recs.pop(position)
+            res.append(item)
+            recommended.add(item.id)
+            if len(topic_recs) == 0:
+                del recs_by_topic[topic]
+
+    if len(res) < count:
+        res.extend(r for r in recs if r.id not in recommended)
```

Comment:
These recs should also be spread by topic, because otherwise people who choose to skip topic selection might all get 3 recs from the same topic.

Here's a test case:
```python
def test_no_preferred_topics(self):
    topics, recs = self._prepare_recs()
    result = rank_by_preferred_topics(recs, [], count=3)
    assert len({r.topic for r in result}) == 3
```

Reply:
This is a good idea. I didn't consider this use case because I thought we force a user to choose at least one topic (this is what many products do).